torchdata/datapipes/iter/util/prefetcher.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import threading
+import time
+
+from typing import Optional
+
+from torchdata.dataloader2 import communication
+
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+PRODUCER_SLEEP_INTERVAL = 0.0001  # Interval between buffer fullfilment checks
+CONSUMER_SLEEP_INTERVAL = 0.0001  # Interval between checking items availablitity in buffer
+
+
+class _PrefetchData:
+    def __init__(self, source_datapipe, buffer_size):
+        self.run_prefetcher = True
+        # TODO: Potential optimization is changing buffer from list to dequeue
+        self.prefetch_buffer = []
+        self.buffer_size = buffer_size
+        self.source_datapipe = source_datapipe
+
+
+@functional_datapipe("prefetch")
+class PrefetcherIterDataPipe(IterDataPipe):
+    def __init__(self, source_datapipe, buffer_size=10):
+        self.source_datapipe = source_datapipe
+        self.buffer_size = buffer_size
+        self.thread: Optional[threading.Thread] = None
+
+    @staticmethod
+    def thread_worker(prefetch_data):
+        itr = iter(prefetch_data.source_datapipe)
+        stop_iteration = False
+        while prefetch_data.run_prefetcher:
+            if len(prefetch_data.prefetch_buffer) < prefetch_data.buffer_size and not stop_iteration:
+                try:
+                    item = next(itr)
+                    prefetch_data.prefetch_buffer.append(item)
+                except StopIteration:
+                    stop_iteration = True
+                except communication.iter.InvalidStateResetRequired:
+                    stop_iteration = True
+                except communication.iter.TerminateRequired:
+                    prefetch_data.run_prefetcher = False
```

Comment:
Do we need to handle those two exceptions? communication is more or less a sub module for MultiprocessingReadingService. I personally feel better to remove those exceptions from a DataPipe

Reply:
I have to it here, because it is a separate thread that needs to be terminated nicely in case of source Datapipe is out of commission.

Reply:
Oh, I see. This prefetch is not only attached to the main problem but also to child processes.